lib/errors/index.js

```diff
@@ -226,21 +226,118 @@ exports.UnknownConstraintError = UnknownConstraintError;
+
+    this.instance       = inst || null;
+
+    this.validatorKey   = validatorKey || null;
+    this.validatorName  = fnName || null;
+    this.validatorArgs  = fnArgs || [];
+
+    if (type) {
+      if (ValidationErrorItem.USE_OLD_TYPES || ValidationErrorItem.TYPES[ type ]) {
+        this.type = type;
+      } else {
+        const lowercaseType = (type + '').toLowerCase().trim();
+        const realType  = ValidationErrorItem.TypeStringMap[ lowercaseType ];
+
+        if (realType && ValidationErrorItem.TYPES[ realType ]) {
+          this.type           = realType;
+          this.originalType   = type; // expose the value of "type" as it was in previous version (just in case someone is reliant on it?)
+        }
+      }
+    }
+  }
+
+  /**
+   * return a lowercase, trimmed string "key" that identifies the validator.
+   *
+   * N.B. the string will be empty if the instance has neither a valid `validatorKey` property nor a valid `validatorName` property
+   *
+   * @param   {Boolean} useTypeAsNS   - optional, defaults to TRUE, controls whether the returned value is "namespace",
+   *                                    this parameter is ignored if the validator's `type` is not one of ValidationErrorItem.TYPES
+   * @param   {String}  NSSeparator   - a separator string for concatenating the namespace, must be not be empty,
+   *                                    defaults to "." (fullstop). only used and validated if useTypeAsNS is TRUE.
+   * @throws  {Error}                 - thrown if NSSeparator is found to be invalid.
+   * @return  {String}
+   */
+  getValidatorKey(useTypeAsNS/* = true */, NSSeparator/* = '.'*/) {
+    // parameter defaults: some CI instances give a syntax error on actual parameter default values
+    const
+      useTANS = typeof useTypeAsNS === 'undefined' ?  true : !!useTypeAsNS,
+      NSSep   = typeof NSSeparator === 'undefined' ? '.' : NSSeparator;
+
+    const
+      type  = this.type,
+      key   = this.validatorKey || this.validatorName,
+      useNS = useTANS && type && ValidationErrorItem.TYPES[ type ];
+
+    if (useNS && (typeof NSSep !== 'string' || !NSSep.length)) {
+      throw new Error('Invalid namespace separator given, must be a non-empty string');
+    }
+
+    if (!(typeof key === 'string' && key.length)) {
+      return '';
+    }
+
+    return (useNS ? [type, key].join(NSSep) : key).trim().toLowerCase();
+  }
+
+
+  /**
+   * generate custom output for JSON.stringify(),
+   * this avoids errors related to serializing circular-refs (due to the `this.instance` property)
+   *
+   * @return  {String}
+   */
+  toJSON() {
+    const o = {
+      message       : this.message,
+      type          : this.type,
+      path          : this.path,
+      value         : this.value,
+      validatorKey  : this.validatorKey,
+      validatorName : this.validatorName,
+      validatorArgs : this.validatorArgs,
+      instance      : this.instance ? this.instance.get({ plain: true }) : null
```

Comment:
I knew about the `Model.toJSON()` but I was not sure as to it's behaviour regarding "includes" ... I wanted to explicitly use the `{ plain: true }` option. 

you are right about `JSON.stringify()` handling the recursion correctly.

what would you like to happen here? should I remove this `toJSON()` method?